lib/spack/spack/extensions.py

```diff
@@ -50,14 +50,47 @@ def load_command_extension(command, path):
     if not extension:
         return None
 
+    # Compute the name of the module we search, exit early if already imported
+    cmd_package = '{0}.{1}.cmd'.format(__name__, extension)
+    python_name = command.replace('-', '_')
+    module_name = '{0}.{1}'.format(cmd_package, python_name)
+    if module_name in sys.modules:
+        return sys.modules[module_name]
+
+    def ensure_package_creation(name):
+        package_name = '{0}.{1}'.format(__name__, name)
+        if package_name in sys.modules:
+            return
+
+        parts = [path] + name.split('.') + ['__init__.py']
+        init_file = os.path.join(*parts)
+        if os.path.exists(init_file):
+            m = llnl.util.lang.load_module_from_file(package_name, init_file)
+        else:
+            m = types.ModuleType(package_name)
+
+        # Setting __path__ to give spack extensions the
+        # ability to import from their own tree, see:
+        #
+        # https://docs.python.org/3/reference/import.html#package-path-rules
+        #
+        m.__path__ = [os.path.dirname(init_file)]
+        sys.modules[package_name] = m
+
+    # Create a searchable package for both the root folder of the extension
+    # and the subfolder containing the commands
+    ensure_package_creation(extension)
+    ensure_package_creation(extension + '.cmd')
+
     # Compute the absolute path of the file to be loaded, along with the
     # name of the python module where it will be stored
     cmd_path = os.path.join(path, extension, 'cmd', command + '.py')
-    python_name = command.replace('-', '_')
-    module_name = '{0}.{1}'.format(__name__, python_name)
 
     try:
+        # TODO: Upon removal of support for Python 2.6 substitute the call
```

Comment:
@alalazo sorry - another question: does this apply to all calls to `load_module_from_file`, or just this one? If it's just this one, then why is that?

Reply:
It's just this one. The reason has to do with how the import mechanism works (I had to read carefully through the docs to understand that). This snippet:

https://github.com/spack/spack/blob/cdfcf7aa76500f8ac519a31d2dc576f0ccda4226/lib/spack/spack/extensions.py#L72-L78

is used to add a search `__path__` to both the "root" extension module and its  command submodule, e.g. `spack.extensions.testcommand` and `spack.extensions.testcommand.cmd`, which are imported here:

https://github.com/spack/spack/blob/cdfcf7aa76500f8ac519a31d2dc576f0ccda4226/lib/spack/spack/extensions.py#L80-L83

After we do this, we won't need the absolute path of each command file anymore, as that will be computed from the `__path__` of the corresponding subpackage.  For instance, if I import `spack.extensions.testcommand.cmd.hello` the import mechanism will compute the corresponding file path from the first `__path__` attribute found going up with subpackages i.e. `spack.extensions.testcommand.cmd`.

Now, in python 2.6 to import a module based on a computed string we need either to use directly the `__import__` function or the `imp` module - and both approaches are not recommended. From python 2.7 on there's `importlib.import_module` which seems to be the idiomatic way and the comment above is a reminder that we could just use that instead of loading the module and adding it to `sys.modules`.

Reply:
@scheibelp Let me know if something should be clarified in the comment above or if it's fine like that. Thanks!